pkg/image/docker/pull.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+ref: https://github.com/kubernetes-sigs/kind
+*/
+
+package docker
+
+import (
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/heptio/sonobuoy/pkg/image/exec"
+)
+
+// PullIfNotPresent will pull an image if it is not present locally
+// retrying up to retries times
+// returns errors from pulling
+func PullIfNotPresent(image string, retries int) error {
+	// TODO(bentheelder): switch most (all) of the logging here to debug level
+	// once we have configurable log levels
+	// if this did not return an error, then the image exists locally
+	cmd := exec.Command("docker", "inspect", "--type=image", image)
+	if err := cmd.Run(); err == nil {
+		log.Infof("Image: %s present locally", image)
+		return nil
+	}
+	// otherwise try to pull it
+	return Pull(image, retries)
+}
+
+// Pull pulls an image, retrying up to retries times
+func Pull(image string, retries int) error {
+	log.Infof("Pulling image: %s ...", image)
+	err := exec.Command("docker", "pull", image).Run()
+	// retry pulling up to retries times if necessary
```

Comment:
You have this cmder + loop logic a number of times. I'd suggest extracting the "loop a cmd with retry" logic to a single place and just call each time with the cmder (or args and make the cmder in that func)

Reply:
It is interfaced out in the `exec` package. You think we need a further abstraction on top of that interface?

Reply:
I'm just saying that we're just repeating ourselves if in push/pull/tag are writing the same loop/retry/backoff logic and you could DRY that up a bit if you put the loop into a single method and just call `execWithRetries` with your args (or something like that)